src/passes/FloatTrap.cpp

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "passes/FloatTrap.h"
+
+#include "asm_v_wasm.h"
+#include "mixed_arena.h"
+#include "wasm-builder.h"
+#include "wasm-printing.h"
+#include "wasm-type.h"
+#include "asmjs/shared-constants.h"
+
+namespace wasm {
+
+Name I64S_REM("i64s-rem"),
+     I64U_REM("i64u-rem"),
+     I64S_DIV("i64s-div"),
+     I64U_DIV("i64u-div");
+
+FloatTrap::FloatTrap(Mode mode)
+    : mode(mode),
+      didAddF64ToI64JSImport(false) {
+  name = "binaryen-trap-mode";
+}
+
+void FloatTrap::visitUnary(Unary* curr) {
+  replaceCurrent(makeTrappingUnary(curr));
+}
+
+void FloatTrap::visitBinary(Binary* curr) {
+  replaceCurrent(makeTrappingBinary(curr));
+}
+
+void FloatTrap::visitModule(Module* curr) {
+  for (const auto& pair : generatedFunctions) {
+    curr->addFunction(pair.second);
+  }
+  generatedFunctions.clear();
+}
+
+Expression* FloatTrap::makeTrappingBinary(Binary* curr) {
+  Name name = getBinaryFuncName(curr);
+  if (mode == Mode::Allow || !name.is()) {
+    return curr;
+  }
+
+  // the wasm operation might trap if done over 0, so generate a safe call
+  WasmType type = curr->type;
+  MixedArena& allocator(getModule()->allocator);
+  auto *call = allocator.alloc<Call>();
+  call->target = name;
+  call->operands.push_back(curr->left);
+  call->operands.push_back(curr->right);
+  call->type = type;
+  ensureBinaryFunc(curr);
+  return call;
+}
+
+Expression* FloatTrap::makeTrappingUnary(Unary* curr) {
+  Name name = getUnaryFuncName(curr);
+  if (mode == Mode::Allow || !name.is()) {
+    return curr;
+  }
+
+  Module& wasm = *getModule();
+  MixedArena& allocator(wasm.allocator);
+  // WebAssembly traps on float-to-int overflows, but asm.js wouldn't, so we must do something
+  // We can handle this in one of two ways: clamping, which is fast, or JS, which
+  // is precisely like JS but in order to do that we do a slow ffi
+  // If i64, there is no "JS" way to handle this, as no i64s in JS, so always clamp if we don't allow traps
+  if (curr->type != i64 && mode == Mode::JS) {
+    // WebAssembly traps on float-to-int overflows, but asm.js wouldn't, so we must emulate that
```

Comment:
Nit: We aren't really "emulating" it at all, we are literally actually calling JS to do the conversion.

Reply:
Almost all of the comments were just grabbed as-is from asm2wasm

... which makes references to asm.js all the more misleading.

And "handle" is probably the better word here?